eth/api_backend.go

```diff
@@ -343,3 +345,45 @@ func (b *EthAPIBackend) StatesInRange(ctx context.Context, fromBlock *types.Bloc
 func (b *EthAPIBackend) StateAtTransaction(ctx context.Context, block *types.Block, txIndex int, reexec uint64) (core.Message, vm.BlockContext, *state.StateDB, func(), error) {
 	return b.eth.stateAtTransaction(block, txIndex, reexec)
 }
+
+func (b *EthAPIBackend) AccessList(ctx context.Context, block *types.Block, reexec uint64, args *ethapi.SendTxArgs) (*types.AccessList, error) {
+	if block == nil {
+		block = b.CurrentBlock()
+	}
+	db, release, err := b.eth.stateAtBlock(block, reexec)
+	defer release()
+	if err != nil {
+		return nil, err
+	}
+	args.SetDefaults(ctx, b)
+	var (
+		gas        uint64
+		accessList = args.AccessList
+		msg        types.Message
+	)
+	for i := 0; i < 10; i++ {
+		// Copy the original db so we don't modify it
+		statedb := db.Copy()
+		// If we have an accesslist, use it
+		if accessList != nil {
+			msg = types.NewMessage(args.From, args.To, uint64(*args.Nonce), args.Value.ToInt(), uint64(*args.Gas), args.GasPrice.ToInt(), *args.Data, *accessList, false)
+		} else {
+			msg = types.NewMessage(args.From, args.To, uint64(*args.Nonce), args.Value.ToInt(), uint64(*args.Gas), args.GasPrice.ToInt(), *args.Data, nil, false)
+		}
+		// Apply the transaction
+		context := core.NewEVMBlockContext(block.Header(), b.eth.blockchain, nil)
+		vmenv := vm.NewEVM(context, core.NewEVMTxContext(msg), statedb, b.eth.blockchain.Config(), vm.Config{})
+		res, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply transaction: %v err: %v", args.ToTransaction().Hash(), err)
+		}
+		if res.UsedGas == gas {
+			vmenv.StateDB.UnprepareAccessList(args.From, args.To, vmenv.ActivePrecompiles())
```

Comment:
I don't see why we need this Unprepare, whicih also is journalled.

Reply:
Yeah, that was a question I had, whether I should unprepare the access list inside or outside of the statedb.
If I unprepare inside of the statedb (the way it currently is) then this should be journalled. But I guess I should unprepare it outside of the statedb.
We need unprepare to remove sender, recipient and the precompiles from the access list
Note to myself, we should only remove the recipient if it has no other state accesses

Reply:
> Note to myself, we should only remove the recipient if it has no other state accesses

Yes. only if the slotlist is empty.

Reply:
But really, you only need to 'clean it' as the very last step before returning it to the user.